vent/coordinator/control_module.py

```diff
@@ -255,28 +295,84 @@ def get_sensors_values(self):
         return self.sensor_values
 
 
+    def update_alarms(self):
+        ''' This goes through the LAST waveform, and updates alarms.'''
+        this_cycle = self.Controller.cycle_counter
+
+        if this_cycle > 1:   #The first cycle for which we can calculate this is cycle "1".
+            data = self.Controller.cycle_waveforms[this_cycle-1]
+            phase = data[:,0]
+            pressure = data[:,1]
+
+            # get the pressure niveau heuristically (much faster than fitting)
+            # 20 and 80 percentiles pulled out of my hat.
+            PEEP = np.percentile(pressure,20)
+            PIP  = np.percentile(pressure,80)
+
+            # measure time of reaching PIP, and leaving PIP
+            last_PIP   = phase[ np.max(np.where(pressure>PIP)) ]
+            first_PIP  = phase[ np.min(np.where(pressure>PIP)) ]
+            # and measure the same for PEEP
+            first_PEEP = phase[ np.min(np.where(np.logical_and(pressure<PEEP, phase > 1))) ]
+            last_PEEP  = phase[-1]
+
+            bpm     = 60./last_PEEP  # 60 sec divided by the duration of last waveform
+            I_phase = last_PIP       # last time-point of the plateau
+
+            self.alarms = test_critical_levels(min = self.PIP_min,      max = self.PIP_max,      value = PIP,       name = "PIP",                alarms = self.alarms)
+            self.alarms = test_critical_levels(min = self.PIP_time_min, max = self.PIP_time_max, value = first_PIP, name = "PIP_TIME",           alarms = self.alarms)
+            self.alarms = test_critical_levels(min = self.PEEP_min,     max = self.PEEP_max,     value = PEEP,      name = "PEEP",               alarms = self.alarms)
+            self.alarms = test_critical_levels(min = self.bpm_min,      max = self.bpm_max,      value = bpm,       name = "BREATHS_PER_MINUTE", alarms = self.alarms)
+            self.alarms = test_critical_levels(min = self.I_phase_min,  max = self.I_phase_max,  value = I_phase,   name = "I_PHASE",            alarms = self.alarms)
+
     def get_alarms(self):
-        return 1
+        #Returns all alarms
+        return self.alarms
+
+    def get_active_alarms(self):
+        #Returns only the active alarms
+        active_alarms = [alarm for alarm in self.alarms if alarm.is_active]
```

Comment:
One potential suggestion, we could have a dictionary of active alarms and a list of logged alarms. This would reduce some of the looping. To check a min / max value against current alarm state we just look in the dictionary. To move an active alarm to logged (after it returns to normal state) we just set the dictionary value to None and move the values to the logged array.

Reply:
...and done as well. Is this what you had in mind, Grant?

Reply:
Yes, looks good. One open question, for get_active_alarms(), do we want to create a list of them from the dictionary or just return the dictionary?

Reply:
While a list might be more consistent with the other return types, I felt like a dictionary would be more versatile for code downstream. This was a conscious choice.

Do you have a preference?

Reply:
Your choice makes sense to me also.